more robust service creation

classicist opened the pull request:
Hi,
    Thank you for writing this great gem. I needed to extend the RPC functionality for my distributed team and thought the code might be more generally useful. The changes I made only extend the functionally you had coded already, so it will not break backward compatibility. Please let me know if you think these changes would be more generally useful. Thanks again for your excellent work -- yours is the clear winner as far as protobuf gems go.

Commit Msg: added more robust service creation with type-safe responses, injected service handlers, and a service registry to faciltate pure-ruby RPC without coupling it to any particular communication strategy (TCPSocket, RabbitMQ, etc) or application architecture.

Commit:
added more robust service creation with type-safe responses, injected service handlers, and a service registry to faciltate pure-ruby RPC without coupling it to any particular communication strategy (TCPSocket, RabbitMQ, etc) or application architecture

Commit:
included the name of the class on which a message field is invalid in the exception raised for an invalid field existing during serialization

Commit:
explictly required require compiler/fully_qualified_name in runtime/service

Commit:
another go at sorting out the load path

Commit:
another go at sorting out the load path

Commit:
fixed problem of nil invalidly overwriting default message constructor value

Commit:
removed hashification of messages passed to rpcs

Commit:
removed generated request type checking

bufdev commented on lib/protocol_buffers.rb:
```diff
@@ -4,13 +4,15 @@ module ProtocolBuffers
   # for 1.9.2 compatibility
   def self.bin_sio(*args)
     sio = StringIO.new(*args)
-    sio.set_encoding('binary') if 
```
whitespace

bufdev commented on lib/protocol_buffers.rb:
```diff
@@ -4,13 +4,15 @@ module ProtocolBuffers
   end
 end
 
+$LOAD_PATH << File.dirname(__FILE__)
```
whitespace

bufdev commented on lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:
```diff
@@ -46,7 +47,7 @@ def write(io)
         dump_service(descriptor.package, service)
       end
     end
-    
```
whitespace

bufdev commented on lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:
```diff
@@ -77,11 +78,11 @@ def line(str = nil)
     end
     @io.write("\n")
   end
-  
```
whitespace

bufdev commented on lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:
```diff
@@ -77,11 +78,11 @@ def line(str = nil)
-  
+
   def in_namespace(type, namespace, rest = "")
-    
```
whitespace

bufdev commented on lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:
```diff
@@ -170,13 +171,21 @@ def dump_enum(package, enum)
     line
   end
 
-  def dump_service(package, service)
```
is this correct indentation?

bufdev commented on lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:
```diff
@@ -170,13 +171,21 @@ def dump_enum(package, enum)
       line
       service.method.each do |method|
         line %{rpc :#{underscore(method.name)}, "#{method.name}", #{service_typename(method.input_type)}, #{service_typename(method.output_type)}}
+        unless fully_qualified_name.nil?
```
silent error

bufdev commented on lib/protocol_buffers/compiler/fully_qualified_name.rb:
```diff
@@ -0,0 +1,35 @@
+      service_typename(fully_qualified_name.to_s).split('::').inject(Object) { |mod, class_name| mod.const_get(class_name) if mod.const_defined?(class_name)}
+    end
+
+    def self.service_typename(type_name)
```
I dont think this kind of stuff should be exposed outside the compiler

bufdev commented on lib/protocol_buffers/compiler/fully_qualified_name.rb:
```diff
@@ -0,0 +1,35 @@
+      lower_case_and_underscored_word.to_s.gsub(/(?:^|_)(.)/) { $1.upcase }
+    end
+
+    def self.underscore(camelized_word)
```
Same comment on exposing outside the compiler

bufdev commented on lib/protocol_buffers/compiler/fully_qualified_name.rb:
```diff
@@ -0,0 +1,35 @@
+module ProtocolBuffers
+  class FullyQualifiedName
+    def self.to_class(fully_qualified_name)
+      return nil if fully_qualified_name.nil?
```
silent error

bufdev commented on lib/protocol_buffers/runtime/message.rb:
```diff
@@ -231,7 +231,8 @@ class Message
     #   message = MyMessageClass.new
     #   message.attributes = attributes
-    def initialize(attributes = {})
+    def initialize(attributes = nil)
```
revert

bufdev commented on lib/protocol_buffers/runtime/message.rb:
```diff
@@ -513,7 +514,7 @@ def self.valid?(message, raise_exception=false)
         next if message.value_for_tag?(tag) && (field.class != Field::MessageField || message.value_for_tag(tag).valid?)
         return false unless raise_exception
-        raise(ProtocolBuffers::EncodeError.new(field), "Required field '#{field.name}' is invalid")
+        raise(ProtocolBuffers::EncodeError.new(field), "Required field '#{field.name}' for '#{self}' is invalid")
```
revert

bufdev commented on lib/protocol_buffers/runtime/message.rb:
```diff
@@ -338,7 +339,7 @@ def ==(obj)
       end
       return true
     end
-    
```
whitespace

bufdev commented on lib/protocol_buffers/runtime/service.rb:
```diff
@@ -1,10 +1,60 @@
+require 'protocol_buffers/compiler/fully_qualified_name'
+
+module ProtocolBuffers
+  class ServiceRegistry
```
document if this is going to stay

bufdev commented on lib/protocol_buffers/runtime/service.rb:
```diff
@@ -1,10 +1,60 @@
+module ProtocolBuffers
+  class ServiceRegistry
+    def self.register_service(fully_qualified_name, handler_factory)
+      return nil unless fully_qualified_name && handler_factory
```
silent error on fully_qualified_name

bufdev commented on lib/protocol_buffers/runtime/service.rb:
```diff
@@ -1,10 +1,60 @@
+  class ServiceRegistry
+    def self.register_service(fully_qualified_name, handler_factory)
+      return nil unless fully_qualified_name && handler_factory
+      @handler_factory_registry ||= Hash.new
```
In your tests, you don't subclass ServiceRegistry, which effectively means you are treating this as a singleton. This means if someone does subclass this, your instance variables (@) will result in two different set of values for @handler_factory_registry.

This is also not thread-safe but I don't think that's a requirement overall.

bufdev commented on lib/protocol_buffers/runtime/service.rb:
```diff
@@ -1,10 +1,60 @@
+    end
+
+    def ensure_correct_request_type!(method_name, message)
+      raise "Request Type Error: #{message} must be an instance of #{types_for(method_name)[:request]}" unless message.instance_of?(types_for(method_name)[:request])
```
specify what type of StandardError you want to raise?

bufdev commented on lib/protocol_buffers/runtime/service.rb:
```diff
@@ -1,10 +1,60 @@
+  end
+end
 
 module ProtocolBuffers
```
This needs documentation now that it is more complicated.

bufdev commented on spec/proto_files/services.pb.rb:
```diff
@@ -38,13 +38,41 @@ class FooBarService < ::ProtocolBuffers::Service
+
   end
   class NoNameFooBarService < ::ProtocolBuffers::Service
-    # purposefully removing qualified name to make sure that nothing breaks
```
whitepace

bufdev commented on spec/proto_files/services.pb.rb:
```diff
@@ -38,13 +38,41 @@ class FooBarService < ::ProtocolBuffers::Service
+    end
+
   end
-end
```
whitespace

Commit:
added a from_hash method to allow contruction of message types with nested AggregateFields from a set of nested hashes

Commit:
added copy constructor for ::ProtocolBuffers::Message in to_hash

Commit:
fixed bug on message class type inspection

Commit:
added recursive functionality to Message.from_hash

Commit:
added support for repeated fields in from_hash

Commit:
removed needless skrewing with the

Commit:
added built gem file

Commit:
added *args to service methods and delegation to handlers

Commit:
added ablility to check whether a default value was set on a field in the proto files without setting it in ruby; added default values set in proto files to the output of to_hash

Commit:
WIP: spiked added defaults to equality for messages

Commit:
Don't populate field when field and value are not specified in the hash

Commit:
Ignore vendor/bundle, .bundle

Commit:
Add ruby-version file

Commit:
Use pry-byebug, debugger isn't for ruby 2

Commit:
Update to rspec 3.x

Commit:
Update to rspec 3.x expect syntax

Commit:
Update ruby-version file for newest ruby

Commit:
respect nil values for imbedded objects

* from_hash was synthesizing an empty object
  instead of setting the property to nil when
  a nil value was passed

* now you really can set a property to nil

Commit:
Update to ruby 2.2.2

Commit:
Force rake to use < 11 due to last_comment removal

emschwar committed:
Allow Enums to be initialized with strings

Commit:
Update ruby-version to 2.2.8 for CI

codekitchen commented:
Hi @emschwar , @jgeiger are you looking to revive this PR? I'm not really sure what to make of your recent pushes, but it's been years so it'd take me a while to refresh my cache on what this is all about. Please let me know if you plan to pursue this. Thanks.

Commit:
Raise more meaningful error on unknown fully_qualified_name

This raises a meaningful error instead of NoMethodError: undefined method const_defined? on NilClass

jacobthemyth committed:
Merge pull request #1 from nexiahome/improve_error

Raise more meaningful error on unknown fully_qualified_name

Commit:
fix: Update .ruby-version for CI

This will allow CI to build the application as we're not really using
specific Ruby code. We're also running the apps using this gem on the
Ruby 2.5.x branch.

[IOTPLAT-710]

Commit:
Update .ruby-version for CI

This will allow CI to build the application.

jgeiger-trane committed:
fix: Cleanup .gitignore and .rspec for CI

Testing locally means installing to vendor which should be ignored so we
don't commit it.

rspec changes the format to print dots instead of the full text.

[IOTPLAT-710]

codekitchen commented:
closing at request of @jgeiger-trane
